### ISSUE TITLE
Old changeSet is executed before legacy migration

### ISSUE THREAD

**Issue by kapmati** [mongock/mongock#452]
## Description

Hi, I was trying to migrate from Mongobee to Mongock and I found some unexpected behavior. It looks like Mongock executes change set before migrating this change set from legacy Mongobee collection. So it is executed once again. I was expecting that Mongock will migrate old collection and will not execute old change set. I didn't find any properties to change this behavior.

Take a look at **executionId** field, it shows that legacy migration was executed AFTER Mongock found and executed old migration. Details below. I prepared a demo spring application, so you can reproduce this behavior easily.

## PRIORITY

NORMAL

### Mongock
- Mongock version: **5.0.29**
- Modules involved: **springboot, mongodb**
- How Mongock is used: **annotation**

### Environment
- spring-boot 2.6.1
- spring-boot-starter-data-mongodb 2.6.1
- spring-data-mongodb 3.3.0
- org.mongodb:mongodb-driver-sync 4.4.0

## Steps to Reproduce
1. Create a java class with a change set.
2. Use mongobee to run this change set or create mongo collection **changelog** with document:
```
{
	"_id" : ObjectId("61b1cca4dcf99872febea207"),
	"changeId" : "DESCRIPTION",
	"author" : "AUTHOR",
	"timestamp" : ISODate("2021-12-09T10:30:12.765+01:00"),
	"changeLogClass" : "com.kapmati.mongomigration.migration.DbChangeLog",
	"changeSetMethod" : "oldMigrationDesc"
}
```
3. Run application with **legacy migration** feature to migrate old changelog

## Behaviour

**Expected behavior:** Old change set is migrated to Mongock mongo collection but is not executed again.
mongockChangeLog collection:
```
/* 1 createdAt:12/9/2021, 2:58:17 PM*/
{
	"_id" : ObjectId("61b20b79e02ce021e2be0304"),
	"author" : "mongock",
	"changeId" : "mongock-legacy-migration",
	"executionId" : "2021-12-09T14:58:17.059826-872",
	"changeLogClass" : "io.mongock.driver.mongodb.sync.v4.changelogs.runonce.MongockSync4LegacyMigrationChangeLog",
	"changeSetMethod" : "mongockSpringLegacyMigration",
	"errorTrace" : null,
	"executionHostname" : "fake-host",
	"executionMillis" : NumberLong(25),
	"metadata" : null,
	"state" : "EXECUTED",
	"timestamp" : ISODate("2021-12-09T14:58:17.130+01:00"),
	"type" : "EXECUTION"
},

/* 2 createdAt:12/9/2021, 2:58:17 PM*/
{
	"_id" : ObjectId("61b20b79e02ce021e2be0302"),
	"author" : "AUTHOR",
	"changeId" : "DESCRIPTION",
	"executionId" : "legacy_migration-2021-12-09T14:58:17.123662-606",
	"changeLogClass" : "com.kapmati.mongomigration.migration.DbChangeLog",
	"changeSetMethod" : "oldMigrationDesc",
	"errorTrace" : null,
	"executionHostname" : "unknown",
	"executionMillis" : NumberLong(-1),
	"metadata" : {
		"migration-type" : "legacy"
	},
	"state" : "EXECUTED",
	"timestamp" : ISODate("2021-12-09T10:30:12.765+01:00"),
	"type" : "EXECUTION"
}
```

**Actual behavior:** Old change set is migrated to Mongock mongo collection and is executed again.
mongockChangeLog collection:
```
/* 1 createdAt:12/9/2021, 2:58:17 PM*/
{
	"_id" : ObjectId("61b20b79e02ce021e2be0306"),
	"author" : "AUTHOR",
	"changeId" : "DESCRIPTION",
	"executionId" : "2021-12-09T14:58:17.059826-872",
	"changeLogClass" : "com.kapmati.mongomigration.migration.DbChangeLog",
	"changeSetMethod" : "oldMigrationDesc",
	"errorTrace" : null,
	"executionHostname" : "fake-host",
	"executionMillis" : NumberLong(1),
	"metadata" : null,
	"state" : "EXECUTED",
	"timestamp" : ISODate("2021-12-09T14:58:17.132+01:00"),
	"type" : "EXECUTION"
},

/* 2 createdAt:12/9/2021, 2:58:17 PM*/
{
	"_id" : ObjectId("61b20b79e02ce021e2be0304"),
	"author" : "mongock",
	"changeId" : "mongock-legacy-migration",
	"executionId" : "2021-12-09T14:58:17.059826-872",
	"changeLogClass" : "io.mongock.driver.mongodb.sync.v4.changelogs.runonce.MongockSync4LegacyMigrationChangeLog",
	"changeSetMethod" : "mongockSpringLegacyMigration",
	"errorTrace" : null,
	"executionHostname" : "fake-host",
	"executionMillis" : NumberLong(25),
	"metadata" : null,
	"state" : "EXECUTED",
	"timestamp" : ISODate("2021-12-09T14:58:17.130+01:00"),
	"type" : "EXECUTION"
},

/* 3 createdAt:12/9/2021, 2:58:17 PM*/
{
	"_id" : ObjectId("61b20b79e02ce021e2be0302"),
	"author" : "AUTHOR",
	"changeId" : "DESCRIPTION",
	"executionId" : "legacy_migration-2021-12-09T14:58:17.123662-606",
	"changeLogClass" : "com.kapmati.mongomigration.migration.DbChangeLog",
	"changeSetMethod" : "oldMigrationDesc",
	"errorTrace" : null,
	"executionHostname" : "unknown",
	"executionMillis" : NumberLong(-1),
	"metadata" : {
		"migration-type" : "legacy"
	},
	"state" : "EXECUTED",
	"timestamp" : ISODate("2021-12-09T10:30:12.765+01:00"),
	"type" : "EXECUTION"
}
```

**How often the bug happens:** Always

## Link to repository using Mongock

https://github.com/kapmati/mongo-migration

**Comment by dieppa** [comment]
Hello @kapmati , your expectation is right. Mongock should migrate the changeLog collection and it shouldn't execute again the changes.

We'll take a look later and spot why that is happening

**Comment by dieppa** [comment]
We have looking at this. We have identified the issue. We will be releasing  a bugfix by 12th Dec

**Comment by dieppa** [comment]
Hello @kapmati ,

Mongock 5.0.31 is released with this bugfix. 

Take into account that if you are using transactions, you will probably need the the transactionManager with high consistency.

Something like this:
```java
@Bean
    public MongoTransactionManager transactionManager(MongoTemplate mongoTemplate) {
        TransactionOptions transactionalOptions = TransactionOptions.builder()
                .readConcern(ReadConcern.MAJORITY)
                .readPreference(ReadPreference.primary())
                .writeConcern(WriteConcern.MAJORITY.withJournal(true))
                .build();
        return new MongoTransactionManager(mongoTemplate.getMongoDbFactory(), transactionalOptions);
    }

```

**Comment by kapmati** [comment]
Hi @dieppa,
thank you for quick response and reaction but i'm afraid that your fix doesn't work. I tested it with `implementation platform("io.mongock:mongock-bom:5.0.31")` version. I don't use transactions in my demo application, but I checked also with new MongoTransactionManager bean with suggested options and it also didn't help. The behavior is still the same.
Could you check check it once again? Maybe 5.0.31 doesn't contain your fix?

**Comment by dieppa** [comment]
Hi @kapmati ,

We are pretty sure it was fixed, but there might be some chance we didn't add the change inn the release, as we were releasing multiple changes.

I will double check later today and will keep you updated.

**Comment by dieppa** [comment]
Hi @kapmati ,

We think the issue is fixed, however your database state is a corrupted state. The changes have been migrated, but, as they were executed, and probably failed, they are marked as failed, so Mongock will try to execute because it thinks they are pending.

We are working inn two approaches:
- Enhancing the legacy migration to cover this scenario(it will fix your case) and we'll probably have it tomorrow or after tomorrow.
- Making a command in the CLI to repair these corrupted states. It will e added to Mongock's road map. Cannot tell you when it will be released


However, inn the meantime, if you cannot wait, you can remove all the migrated changes migrated and the one with id `mongock-legacy-migration` and run Mongock again with legacy migration enabled.

We suggest to wait until we provide an official fix.

**Comment by kapmati** [comment]
Hi @dieppa, I will wait for an official fix. It's not so urgent for me :)

**Comment by dieppa** [comment]
Hi @kapmati ,

this has been fixed in version 5.0.32. The actual mongock library is released and available, however the bom it's still on its way. It will take the time Maven takes to do the artifact publishing.

## Note 
The legacy migration needs to e executed again and I am pretty sure it appears as executed in your mongock collection.
As said, we'll be adding a CLI command `mongock repair legacy-migration`, but it's not there yet and we have other priorities at the moment. So you have two options for this:

-  **(recommended)** Mark your legacy migration as runAlways with `mongock.legacy-migration.run-always=true`. The fact that the legacy migration runs always won't hurt, even if the origin table has been removed
- Manually remove the entry with id `mongock-legacy-migration`

**Comment by kapmati** [comment]
Version 5.0.32 works correctly. Thank you @dieppa
